### ISSUE TITLE
TypeScript: endless 100% CPU when changing many TS files at once

### ISSUE THREAD

**Issue by bpasero** [Microsoft/vscode#64485]
- VSCode Version: insiders
- OS Version: macOS 10.14

Steps to Reproduce:

1. open the `vscode` workspace
2. search for "`copyright`" in `*.ts`
3. replace with "`copyright2`"
4. run and wait

=> the TS extensions seems to go up to a 100% and stays there without ever recovering.

Full command line:

`bpasero          78619 100.1 11.0  6870176 1840020   ??  R     7:14AM   2:13.12 /Applications/Visual Studio Code - Insiders.app/Contents/Frameworks/Code - Insiders Helper.app/Contents/MacOS/Code - Insiders Helper /Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js --useInferredProjectPerProjectRoot --enableTelemetry --cancellationPipeName /var/folders/w_/xgfzkgq14jv_ykk2lm0k_h5c0000gn/T/vscode-typescript/tscancellation-43d76d57848697e4dfca.tmp* --locale en --noGetErrOnBackgroundUpdate`

**Comment by mjbvz** [comment]
I investigated this and have opened https://github.com/Microsoft/TypeScript/issues/29598 to track the issue.

I believe the issue is that TS Server processes the change events sequentially and can end up performing expensive operations for each one. 

Will follow on this issue upstream

**Comment by mjbvz** [comment]
Reopening for VS Code side of the fix

**Comment by mjbvz** [comment]
Expected behavior now: there still should be a bump in cpu usage from the tsserver but it should drop back down after a minute or so.

I still sometimes see bad performance if I have formatOnSave/codeActionsOnSave/autosave turned on. Please open a new issue if you see bad perf when using one of those options

**Comment by bpasero** [comment]
This is not fixed for me, I did the rename of copyright => copyright2 and the CPU is high now for 5 minutes. I had a TS file opened from the workbench, so the language service was warm. I am using `3.4.0-rc` in the status bar.

**Comment by pladaria** [comment]
We have a TS repo with few thousand files, we were getting 100% usage in several cores and the problem was caused by (that was enabled in workspace config):
```
"typescript.tsserver.experimental.enableProjectDiagnostics": true,
```
This makes the eslint extension go crazy constantly checking files. This was barely noticeable on a 16 core machine but puts a 2015 macbook pro on fire.

**Comment by mjbvz** [comment]
Ok, I had a heck of time even testing this scenario because it breaks many other parts of VS Code (and many other extensions). I twice completely corrupted my workspace too.

I tested using `copyright (c)` - `copyright (c2)` in `./src/vs/workbench/**/*.ts` (1500 files).  Make sure to disable all extensions and auto save!

Once I tested this, I see CPU for the main TS server drop after a short time while the cpu usage for the syntax server stays oddly high for two minutes longer. 

The root cause seems to be that for some reason, this ends up opening all the touched files as new  `TextDocuments` which causes them all to be synced over the the TSServer. The syntax server usually only knows about the current file, but now ends up being flooded with 1500 new files to parse. The semantic server on the other hand has already parsed these files

**Comment by mjbvz** [comment]
@jrieken I brought this up at Redmond standup today and @roblourens suggested I loop you in on this issue

When I do a large find/replace in VS Code codebase, I see a `TextDocument` opened for every touched file. This happens even though these files are never opened as user facing editors

This causes some problems for the JS/TS, but also breaks other extensions in my testing. ESlint for example also goes into a high cpu usage loop when this happens. Working around this on the JS/TS side is non-trivial and I'm wondering if this is something we could handle better in core instead

**Comment by roblourens** [comment]
Here is the old discussion: https://github.com/microsoft/vscode/issues/89300

The impact is bad and surprising to extension developers, it would be great if we could open a model in a lightweight way that doesn't get broadcast to the world since we know we will immediately close it. But if extensions want to participate in opening and editing and saving then it might also be surprising to not find out about it.

**Comment by jrieken** [comment]
Hm... This isn't easy. In order to make changes we need to open/create text documents and they get propagated to the extension host by default. Not doing the propagation will yield in various issues I am afraid. We have an iron-hard guarantee that "normal" documents are available on the extension host and various features (in the API) will fail when we don't have that. Like, we might end up with an active text editor that has no text document. 

So, instead of saying the renderer shouldn't sync model the extension host I would rephrase the problem to "the renderer shouldn't open sooo many documents". This might happen in various places but I'd say the bulk edit service is the most likely candidate here. Today, bulk edit works in stages: 

1. open all text documents
2. make checks for all documents (like expected vs actual version id)
3. apply edits to all documents
4. clean-up

This generally makes sense because it gives this some kind of transaction characteristic. However, with huge edits this should be broken up I assume. Also, the undo-redo world might to help so that we can implement this better

**Comment by mjbvz** [comment]
We've made a number of improvements here since this was opened and [TS 7.0 should also dramatically improve performance](https://code.visualstudio.com/updates/v1_107#_typescript-70-preview)

If you still see any perf issues using TypeScript 7.0+ please open an issue with steps to reproduce the problem and I'll take a look
